src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,1121 @@
+            break;
+
+        case CMD_TERMINATE_SESSION:
+            _curr_op = CMD_NONE;
+            _session_valid = false;
+            _stop_timer();
+            _curr_op_result_callback(_translate(_session_result));
+            break;
+
+        case CMD_LIST_DIRECTORY: {
+            bool added = false;
+            uint8_t start = 0;
+            for (uint8_t i = 0; i < payload->size; i++) {
+                if (payload->data[i] == 0) {
+                    std::string entry = std::string(reinterpret_cast<char*>(&payload->data[start]));
+                    if (entry.length() > 0) {
+                        added = true;
+                        _curr_directory_list.emplace_back(entry);
+                    }
+                    start = i + 1;
+                }
+            }
+            if (added) {
+                // Ask for next batch of file names
+                _list_directory(_curr_directory_list.size());
+            } else {
+                // We came to end - report entire list
+                _curr_op = CMD_NONE;
+                _stop_timer();
+                _curr_dir_items_result_callback(MavlinkFTP::Result::SUCCESS, _curr_directory_list);
+            }
+            break;
+        }
+
+        case CMD_CALC_FILE_CRC32: {
+            _curr_op = CMD_NONE;
+            uint32_t checksum = *reinterpret_cast<uint32_t*>(payload->data);
+            _stop_timer();
+            _current_crc32_result_callback(MavlinkFTP::Result::SUCCESS, checksum);
+            break;
+        }
+
+        default:
+            _curr_op = CMD_NONE;
+            _stop_timer();
+            _curr_op_result_callback(MavlinkFTP::Result::SUCCESS);
+            break;
+    }
+}
+
+void MavlinkFTPImpl::_process_nak(PayloadHeader* payload)
+{
+    if (payload != nullptr) {
+        _process_nak(static_cast<ServerResult>(payload->data[0]));
+    }
+}
+
+void MavlinkFTPImpl::_process_nak(ServerResult result)
+{
+    std::lock_guard<std::mutex> lock(_curr_op_mutex);
+    switch (_curr_op) {
+        case CMD_NONE:
+            LogWarn() << "Received NAK without active operation";
+            break;
+
+        case CMD_OPEN_FILE_RO:
+        case CMD_READ_FILE:
+            _session_result = result;
+            if (_session_valid) {
+                _end_read_session();
+            } else {
+                _stop_timer();
+                _curr_op_result_callback(_translate(_session_result));
+            }
+            break;
+
+        case CMD_OPEN_FILE_WO:
+        case CMD_WRITE_FILE:
+            _session_result = result;
+            if (_session_valid) {
+                _end_write_session();
+            } else {
+                _stop_timer();
+                _curr_op_result_callback(_translate(_session_result));
+            }
+            break;
+
+        case CMD_TERMINATE_SESSION:
+            _session_valid = false;
+            _stop_timer();
+            _curr_op_result_callback(_translate(_session_result));
```

Comment:
Can you be sure that the callback is not null?

And you should always use the thread pool for user callbacks:
```
if (_curr_op_result_callback) {
    const auto temp_callback = _curr_op_result_callback;
    _parent->call_user_callback([temp_callback]() { temp_callback(_translate(_session_result)); });
```

And if you make `_translate` `static` you don't need to pass `this` into the lambda.

Reply:
This blew my mind. Very good!